File: src/vct/col/ast/DeclarationStatement.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import java.util.*;

import vct.util.ClassName;

public class DeclarationStatement extends ASTDeclaration {

  // Type should become part of ASTNode!!
  private Type type;
  private ASTNode expr;

  public DeclarationStatement(String name,Type type){
    super(name);
    this.type=type;
    expr=null;
  }

  public DeclarationStatement(String name,Type type,ASTNode expr){
    super(name);
    this.type=type;
    this.expr=expr;
  }
  
  public Type getType() { return type; }

  public String getName() { return name; }
  
  public ASTNode getInit() { return expr; }
  
  @Override
  public ClassName getDeclName() {
    hre.System.Warning("%s.%s",package_name,name);
    return new ClassName(package_name,name);
  }

  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    visitor.visit(this);
  }
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    return map.map(this);
  }

}

File: vercors/src/main/java/vct/col/ast/DeclarationStatement.scala
package vct.col.ast

import hre.lang.System.Debug
import vct.col.util.VisitorHelper
import vct.util.ClassName

/**
 * AST node that represents a declaration statement, for example "`int test := 2+4;`". 
 * 
 * @author sccblom, whmoortwijn
 * @param _name The name of the declared variable, e.g. "`test`".
 * @param _type The type of the declared variable, e.g. "`int`".
 * @param init The expression that determines the initial value of the declared variable, e.g. "`2+4`".
 */
class DeclarationStatement(private[this] val _name:String, private[this] val _type:Type,private[this] val init:ASTNode) extends ASTDeclaration(_name) with VisitorHelper {
  /**
   * Initialises a new AST node that represents a declaration statement without 
   * initial value, for example "`int test;`".
   * 
   * @param name The name of the declared variable, e.g. "`test`".
   * @param t The type of the declared variable, e.g. "`int`".
   */
  def this(name:String, t:Type) = this(name, t, null)
  
  /** Yields the declaration type */
  override def getType() = _type

  def getInit() = init
  def init():ASTNode = init

  /** Yields the full name of the declared variable, including package name. */
  override def getDeclName() = {
    Debug(s"$packageName.$name")
    new ClassName(packageName, name)
  }
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}
